pandas/core/internals/blocks.py

```diff
@@ -1582,6 +1582,50 @@ def _replace_coerce(
                 )
         return self
 
+    def _discriminate_replace_and_coerce_block(
```

Comment:
could we re-use _split_and_operate here?

Reply:
we also do something kinda-similar at the end of Block.where with `for m in [mask, ~mask]`, maybe something can be shared

Reply:
@jbrockmendel would there be performance implications to operating on each column separately (which, as I understand, `split_and_operate` would do)?

Reply:
I would rather take a perf hit (maybe) that adding a ton more code.

Reply:
> would there be performance implications to operating on each column separately (which, as I understand, split_and_operate would do)?

Only a small one.  Building on what @jreback said, the `replace_foo` code is already obscenely complicated, so a small perf hit is better than making that worse.

Reply:
Thanks for the explanation - I've tried writing this using split_and_operate, though it seems a bit complicated - is this kind of what you had in mind @jbrockmendel ?

Currently, this works for:
- any_real_dtype
- datetime64_dtype
- timedelta64_dtype
- complex_dtype

doesn't work for:
- string_dtype
- bytes_dtype
- tz_aware_fixture

Reply:
What I'm struggling with is that method I'd like to apply is a block method, but inside the `f` function in `split_and_operate`, I would only have access to an ndarray

Reply:
> but inside the f function in split_and_operate, I would only have access to an ndarray

Are you sure?  Looking at `Block.fillna` it passes a function `f` to split_and_operate that operates on blocks.

Reply:
@jbrockmendel thanks for pointing me to that, that was useful.

Something I'm stuck on is that if I try writing this as

```python
                    def f(mask, values, idx):
                        if idx is not None:
                            block = b.getitem_block(slice(idx, idx + 1))
                        else:
                            block = b
                        return block._replace_coerce(
                            mask=mask.reshape(block.shape),
                            to_replace=s,
                            value=d,
                            inplace=inplace,
                            convert=convert,
                            regex=regex,
                        )

                    if m.any() or convert:
                        result = b.split_and_operate(m, f, inplace)
                        new_rb = _extend_blocks(result, new_rb)
                    else:
                        new_rb.append(b)
```

then I run into the following problem: if `b` is `ObjectBlock`, then `b.values` is also `ObjectBlock`, which will cause a problem when it's passed to `split_and_operate`:
```python-traceback
(Pdb) !b
ObjectBlock: slice(1, 2, 1), 1 x 4, dtype: object
(Pdb) !b.values
ObjectBlock: slice(1, 2, 1), 1 x 4, dtype: object
(Pdb) !b.values[0]
*** TypeError: 'ObjectBlock' object is not subscriptable
```

This is in contract to when `b` is an `IntBlock`, when it works fine:
```python-traceback
(Pdb) !b
IntBlock: slice(0, 1, 1), 1 x 4, dtype: int64
(Pdb) !b.values
array([[0, 1, 2, 3]])
(Pdb) !b.values[0]
array([0, 1, 2, 3])
```

The code, as it is, would solve issue 30512, but there are problems that arise in other tests because of the above, which I still need to make sense of